intg-denonavr/denon_remote.py

```diff
@@ -0,0 +1,820 @@
+# pylint: disable=C0302
+"""
+Remote entity functions.
+
+:license: Mozilla Public License Version 2.0, see LICENSE for more details.
+"""
+
+from typing import Any
+
+import avr
+import helpers
+import ucapi.remote
+from command_constants import (
+    AudysseyCommands,
+    CoreCommands,
+    DiracCommands,
+    SoundModeCommands,
+    VolumeCommands,
+)
+from config import AvrDevice, create_entity_id
+from media_player import DenonMediaPlayer
+from ucapi import EntityTypes, Remote, StatusCodes, media_player
+from ucapi.remote import Attributes, Commands, Features
+from ucapi.ui import Buttons
+
+# Mapping of an AVR state to a remote entity state
+REMOTE_STATE_MAPPING = {
+    avr.States.ON: ucapi.remote.States.ON,
+    avr.States.OFF: ucapi.remote.States.OFF,
+    avr.States.PAUSED: ucapi.remote.States.ON,
+    avr.States.PLAYING: ucapi.remote.States.ON,
+    avr.States.UNAVAILABLE: ucapi.remote.States.UNAVAILABLE,
+    avr.States.UNKNOWN: ucapi.remote.States.UNKNOWN,
+}
+
+
+# pylint: disable=R0903
+class DenonRemote(Remote):
+    """Representation of a Denon/Marantz AVR Remote entity."""
+
+    def __init__(self, device: AvrDevice, receiver: avr.DenonDevice, denon_media_player: DenonMediaPlayer):
+        """Initialize the class."""
+        self._device: avr.DenonDevice = receiver
+        self._denon_media_player: DenonMediaPlayer = denon_media_player
+        entity_id = create_entity_id(receiver.id, EntityTypes.REMOTE)
+        features = [Features.SEND_CMD, Features.ON_OFF, Features.TOGGLE]
+        super().__init__(
+            entity_id,
+            f"{device.name} Remote",
+            features,
+            attributes={
+                Attributes.STATE: receiver.state,
+            },
+            simple_commands=self._denon_media_player.get_supported_commands(),
+            button_mapping=REMOTE_BUTTONS_MAPPING,
+            ui_pages=DenonRemote._get_remote_ui_pages(),
+        )
+
+    async def command(self, cmd_id: str, params: dict[str, Any] | None = None) -> StatusCodes:
+        """
+        Remote entity command handler.
+
+        Called by the integration-API if a command is sent to a configured remote entity.
+
+        :param cmd_id: command
+        :param params: optional command parameters
+        :return: status code of the command request
+        """
+        if params:
+            repeat = self._get_int_param("repeat", params, 1)
+        else:
+            repeat = 1
```

Comment:
This seems to use the repeat parameter for the entire `send_cmd_sequence`. I have to check what the intention is exactly, to repeat each command in the sequence, or to repeat the sequence. The `repeat` parameter is definitely a confusing parameter for `send_cmd_sequence`...

Reply:
Let me know and I'll adjust it

Reply:
Checked the repeat handling for the infrared-remote entity in the core, and the [remote-entity documentation](https://github.com/unfoldedcircle/core-api/blob/main/doc/entities/entity_remote.md#commands): parameters apply to the individual commands and not to the overall sequence.


cmd_id | Parameters | Type | Description
-- | -- | -- | --
send_cmd_sequence | sequence | String array | Command list. Same defaults are used as for the send_cmd command.
  | repeat | Number | Optional: how many times **each command** shall be repeated.
  | delay | Number | Optional: delay in milliseconds **between commands**.
  | hold | Number | Optional: time in milliseconds before each command is released.

That seems to be the last change, then it's good to merge!